File: 0735-asteroid-collision/0735-asteroid-collision.py
class Solution(object):
    def asteroidCollision(self, asteroids):
        stack = []

        for asteroid in asteroids:
            while stack and asteroid < 0 < stack[-1]:
                if stack[-1] < abs(asteroid):
                    stack.pop()
                    continue
                elif stack[-1] == abs(asteroid):
                    stack.pop()
                break
            else:
                stack.append(asteroid)

        return stack

File: 0735-asteroid-collision/0735-asteroid-collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        stack<int> asteroidStack;

        for (int asteroid : asteroids) {
            bool destroyed = false;
            while (!asteroidStack.empty() && asteroid < 0 && asteroidStack.top() > 0) {
                if (abs(asteroid) > asteroidStack.top()) {
                    asteroidStack.pop();
                } else if (abs(asteroid) == asteroidStack.top()) {
                    asteroidStack.pop();
                    destroyed = true;
                    break;
                } else {
                    destroyed = true;
                    break;
                }
            }

            if (!destroyed) {
                asteroidStack.push(asteroid);
            }
        }

        vector<int> result;
        while (!asteroidStack.empty()) {
            result.insert(result.begin(), asteroidStack.top());
            asteroidStack.pop();
        }

        return result;
    }
};